MacOS 10.15.7, CPython3.9.6: Random pty test failure; close_fds takes 0.6 seconds!

Plan is to defer closing child_fd in parent until *after* first read succeeds. I think this is a race condition related to closing the child_fd in the parent process too quickly -- before the child process has had a chance to make it the controlling tty.

The os.close(child_fd) taking so long is an important clue.


```
_________________________ test_pty_raw_size_inherited __________________________

sh = CmdContext(options=Options(inherit_env=True, input=b'', input_close=False, output=<Redirect.CAPTURE: -10>, output_appe...t_name=None, pass_fds=(), pass_fds_close=False, write_mode=False, start_new_session=False, preexec_fn=None, pty=False))

    @pytest.mark.xfail(_is_uvloop(), reason="uvloop")
    async def test_pty_raw_size_inherited(sh):
        "Test the `pty` option in raw mode."
    
        cmd = sh("stty", "size").set(pty=raw(rows=..., cols=...))
        result = await cmd
    
>       rows, cols = (int(n) for n in result.rstrip().split())
E       ValueError: not enough values to unpack (expected 2, got 0)

tests/unix/test_unix.py:1123: ValueError
------------------------------ Captured log call -------------------------------
WARNING  shellous:pty_util.py:203 _inherit_term_size ex=OSError(25, 'Inappropriate ioctl for device')
INFO     shellous:log.py:91 Runner.__aenter__ stepin <Runner 'stty' pid=None> (macOS-10.15.7 CPython 3.9.7 asyncio.unix_events._UnixSelectorEventLoop MainThread ThreadedChildWatcher)
INFO     shellous:log.py:91 Runner._start stepin <Runner 'stty' pid=None>
INFO     shellous:runner.py:245 _setup_pty1: PtyFds(parent_fd=8, eof=b'', reader=None, writer=None) child_fd=10
INFO     asyncio:base_events.py:1665 execute program 'stty': <_UnixSubprocessTransport pid=9156 running>
WARNING  shellous:log.py:199 close_fds took 0.602129 seconds ex=None
INFO     shellous:log.py:102 Runner._start stepout <Runner 'stty' pid=9156 exit_code=None> ex=None
INFO     shellous:log.py:102 Runner.__aenter__ stepout <Runner 'stty' pid=9156 exit_code=None> ex=None
INFO     shellous:log.py:135 copy_bytearray stepin
WARNING  asyncio:base_events.py:1885 Executing <Task pending name='Task-674' coro=<test_pty_raw_size_inherited() running at /Users/runner/work/shellous/shellous/tests/unix/test_unix.py:1121> wait_for=<Future pending cb=[<TaskWakeupMethWrapper object at 0x10d932160>()] created at /Users/runner/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/asyncio/base_events.py:424> cb=[_run_until_complete_cb() at /Users/runner/hostedtoolcache/Python/3.9.7/x64/lib/python3.9/asyncio/base_events.py:184] created at /Users/runner/Library/Caches/pypoetry/virtualenvs/shellous-fzoxJzh3-py3.9/lib/python3.9/site-packages/pytest_asyncio/plugin.py:193> took 0.605 seconds
INFO     asyncio:unix_events.py:503 <_UnixReadPipeTransport fd=8 polling> was closed by peer
INFO     shellous:log.py:135 write_stream stepin
INFO     shellous:log.py:140 write_stream stepout ex=None
INFO     shellous:pty_util.py:93 PtyStreamReaderProtocol.eof_received
INFO     shellous:pty_util.py:85 PtyStreamReaderProtocol.connection_lost ex=None
INFO     shellous:log.py:140 copy_bytearray stepout ex=None
INFO     shellous:log.py:91 Runner.__aexit__ stepin <Runner 'stty' pid=9156 exit_code=None> exc_value=None
INFO     shellous:log.py:91 Runner._finish stepin <Runner 'stty' pid=9156 exit_code=None>
INFO     shellous:log.py:91 Runner._wait stepin <Runner 'stty' pid=9156 exit_code=None>
INFO     shellous:runner.py:330 waitpid returned (9156, 0)
INFO     shellous:log.py:102 Runner._wait stepout <Runner 'stty' pid=9156 exit_code=0> ex=None
INFO     shellous:log.py:91 Runner._close stepin <Runner 'stty' pid=9156 exit_code=0>
INFO     shellous:pty_util.py:46 PtyFds.close
INFO     shellous:pty_util.py:117 writer_transport.close()
INFO     shellous:log.py:102 Runner._close stepout <Runner 'stty' pid=9156 exit_code=0> ex=None
INFO     shellous:log.py:102 Runner._finish stepout <Runner 'stty' pid=9156 exit_code=0> ex=None
INFO     shellous:log.py:102 Runner.__aexit__ stepout <Runner 'stty' pid=9156 exit_code=0> ex=None exc_value=None

```

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -76,7 +76,7 @@ jobs:
     - name: Run Tests with SafeChildWatcher (Linux and MacOS)
       if: matrix.os == 'ubuntu-latest' || matrix.os == 'macos-latest'
       run: |
-        $POETRY_RUN python -m pytest -vv -s --log-cli-level=$LOG_LEVEL
+        $POETRY_RUN python -m pytest -vv -s --log-cli-level=DEBUG
       env:
         SHELLOUS_CHILDWATCHER_TYPE: "safe"
     - name: Run Tests with MultiLoopChildWatcher (Linux and MacOS)

--- a/shellous/log.py
+++ b/shellous/log.py
@@ -50,7 +50,7 @@ def log_method(enabled, *, _info=False, **kwds):
         is_asyncgen = inspect.isasyncgenfunction(func)
         assert is_asyncgen or inspect.iscoroutinefunction(
             func
-        ), f"Expected {func.__qualname__} to be coroutine or asyncgen"
+        ), f"Expected {func!r} to be coroutine or asyncgen"
 
         if "." in func.__qualname__ and is_asyncgen:
             # Use _asyncgen_wrapper which includes value of `self` arg.

--- a/shellous/pty_util.py
+++ b/shellous/pty_util.py
@@ -16,7 +16,7 @@ try:
 except ImportError:
     pass
 
-from .log import LOGGER, log_method
+from .log import LOG_DETAIL, LOGGER, log_method
 from .util import close_fds
 
 _STDIN_FILENO = 0
@@ -49,7 +49,7 @@ class PtyFds(NamedTuple):
     reader: Any = None
     writer: Any = None
 
-    @log_method(True)
+    @log_method(LOG_DETAIL)
     async def open_streams(self):
         "Open pty reader/writer streams."
         reader, writer = await _open_pty_streams(self.parent_fd, self.child_fd)
@@ -63,7 +63,8 @@ class PtyFds(NamedTuple):
 
     def close(self):
         "Close pty file descriptors."
-        LOGGER.info("PtyFds.close")
+        if LOG_DETAIL:
+            LOGGER.info("PtyFds.close")
         self.child_fd.close()
         if self.writer:
             self.writer.close()
@@ -101,10 +102,10 @@ class PtyStreamReaderProtocol(asyncio.StreamReaderProtocol):
 
     def connection_lost(self, exc):
         "Intercept EIO error and treat it as EOF."
-        LOGGER.info("PtyStreamReaderProtocol.connection_lost ex=%r", exc)
+        if LOG_DETAIL:
+            LOGGER.info("PtyStreamReaderProtocol.connection_lost ex=%r", exc)
         if isinstance(exc, OSError) and exc.errno == errno.EIO:
             exc = None
-            LOGGER.info("connection_lost EIO -> EOF")
         super().connection_lost(exc)
 
     def data_received(self, data):
@@ -116,7 +117,8 @@ class PtyStreamReaderProtocol(asyncio.StreamReaderProtocol):
 
     def eof_received(self):
         "Log when EOF received."
-        LOGGER.info("PtyStreamReaderProtocol.eof_received")
+        if LOG_DETAIL:
+            LOGGER.info("PtyStreamReaderProtocol.eof_received")
         return super().eof_received()
 
 
@@ -145,7 +147,8 @@ async def _open_pty_streams(parent_fd: int, child_fd: ChildFd):
 
     # Patch writer_transport.close so it also closes the reader_transport.
     def _close():
-        LOGGER.info("writer_transport.close()")
+        if LOG_DETAIL:
+            LOGGER.info("writer_transport.close()")
         _orig_close()
         reader_transport.close()
 

--- a/shellous/pty_util.py
+++ b/shellous/pty_util.py
@@ -4,6 +4,7 @@ import asyncio
 import errno
 import os
 import struct
+from dataclasses import dataclass
 from typing import Any, NamedTuple, Optional
 
 # The following modules are not supported on Windows.
@@ -15,7 +16,7 @@ try:
 except ImportError:
     pass
 
-from .log import LOGGER
+from .log import LOGGER, log_method
 from .util import close_fds
 
 _STDIN_FILENO = 0
@@ -24,18 +25,37 @@ _LFLAG = 3
 _CC = 6
 
 
+@dataclass
+class ChildFd:
+    "Make sure child fd is only closed once."
+    child_fd: int
+
+    def __int__(self):
+        return self.child_fd
+
+    def close(self):
+        "Close the child file descriptor exactly once."
+        if self.child_fd >= 0:
+            fdesc = self.child_fd
+            self.child_fd = -1
+            close_fds([fdesc])
+
+
 class PtyFds(NamedTuple):
     "Track parent fd for pty."
     parent_fd: int
+    child_fd: ChildFd
     eof: bytes
     reader: Any = None
     writer: Any = None
 
+    @log_method(True)
     async def open_streams(self):
         "Open pty reader/writer streams."
-        reader, writer = await _open_pty_streams(self.parent_fd)
+        reader, writer = await _open_pty_streams(self.parent_fd, self.child_fd)
         return PtyFds(
             self.parent_fd,
+            self.child_fd,
             self.eof,
             reader,
             writer,
@@ -44,10 +64,11 @@ class PtyFds(NamedTuple):
     def close(self):
         "Close pty file descriptors."
         LOGGER.info("PtyFds.close")
+        self.child_fd.close()
         if self.writer:
             self.writer.close()
         else:
-            os.close(self.parent_fd)
+            close_fds([self.parent_fd])
 
 
 def open_pty(pty_func):
@@ -59,12 +80,10 @@ def open_pty(pty_func):
     if callable(pty_func):
         pty_func(child_fd)
 
-    return (
-        PtyFds(
-            parent_fd,
-            _get_eof(child_fd),
-        ),
-        child_fd,
+    return PtyFds(
+        parent_fd,
+        ChildFd(child_fd),
+        _get_eof(child_fd),
     )
 
 
@@ -88,27 +107,39 @@ class PtyStreamReaderProtocol(asyncio.StreamReaderProtocol):
             LOGGER.info("connection_lost EIO -> EOF")
         super().connection_lost(exc)
 
+    def data_received(self, data):
+        "Close child_fd when first data received."
+        if self._pty_child_fd:
+            self._pty_child_fd.close()
+            self._pty_child_fd = None
+        return super().data_received(data)
+
     def eof_received(self):
         "Log when EOF received."
         LOGGER.info("PtyStreamReaderProtocol.eof_received")
         return super().eof_received()
 
 
-async def _open_pty_streams(file_desc):
+async def _open_pty_streams(parent_fd: int, child_fd: ChildFd):
     "Open reader, writer streams for pty file descriptor."
 
     loop = asyncio.get_running_loop()
     reader = asyncio.StreamReader(loop=loop)
     reader_protocol = PtyStreamReaderProtocol(reader, loop=loop)
+
+    # Stick reference to child_fd into protocol so we can close it after the
+    # first data is received on BSD systems.
+    reader_protocol._pty_child_fd = child_fd
+
     reader_transport, _ = await loop.connect_read_pipe(
         lambda: reader_protocol,
-        os.fdopen(file_desc, "rb", 0, closefd=False),
+        os.fdopen(parent_fd, "rb", 0, closefd=False),
     )
 
     writer_protocol = asyncio.streams.FlowControlMixin()
     writer_transport, writer_protocol = await loop.connect_write_pipe(
         lambda: writer_protocol,
-        os.fdopen(file_desc, "wb", 0, closefd=True),
+        os.fdopen(parent_fd, "wb", 0, closefd=True),
     )
     writer = asyncio.StreamWriter(writer_transport, writer_protocol, reader, loop)
 

--- a/shellous/runner.py
+++ b/shellous/runner.py
@@ -242,10 +242,17 @@ class _RunOptions:
         Initializes `self.pty_fds`.
         """
 
-        self.pty_fds, child_fd = pty_util.open_pty(pty)
-        self.open_fds.append(child_fd)
+        self.pty_fds = pty_util.open_pty(pty)
+        child_fd = int(self.pty_fds.child_fd)
 
-        LOGGER.info("_setup_pty1: %r child_fd=%r", self.pty_fds, child_fd)
+        # On BSD-derived systems like FreeBSD and Darwin, we delay closing the
+        # pty's child_fd in the parent process until after the first read
+        # succeeds. On Linux, we close the child_fd as soon as possible.
+
+        if not _BSD:
+            self.open_fds.append(self.pty_fds.child_fd)
+
+        LOGGER.info("_setup_pty1: %r", self.pty_fds)
 
         if stdin == asyncio.subprocess.PIPE:
             stdin = child_fd

--- a/shellous/runner.py
+++ b/shellous/runner.py
@@ -79,7 +79,7 @@ class _RunOptions:
         "Make sure those file descriptors are cleaned up."
         self.close_fds()
         if exc_value:
-            LOGGER.info(
+            LOGGER.warning(
                 "_RunOptions.exit %r exc_value=%r", self.command.name, exc_value
             )
 
@@ -252,7 +252,8 @@ class _RunOptions:
         if not _BSD:
             self.open_fds.append(self.pty_fds.child_fd)
 
-        LOGGER.info("_setup_pty1: %r", self.pty_fds)
+        if LOG_DETAIL:
+            LOGGER.info("_setup_pty1: %r", self.pty_fds)
 
         if stdin == asyncio.subprocess.PIPE:
             stdin = child_fd
@@ -351,7 +352,8 @@ class Runner:
 
         while True:
             pid, status = os.waitpid(proc_pid, os.WNOHANG)
-            LOGGER.info("waitpid returned %r", (pid, status))
+            if LOG_DETAIL:
+                LOGGER.info("waitpid returned %r", (pid, status))
 
             if pid == proc_pid:
                 self.proc._transport._returncode = status

